core/arch/arm/kernel/elf_load.c

```diff
@@ -326,6 +343,10 @@ static TEE_Result load_head(struct elf_load_state *state, size_t head_size)
 		n--;
 		copy_phdr(&phdr, state, n);
 	} while (phdr.p_type != PT_LOAD);
+
+	/* Check for integer overflow */
+	if (phdr.p_vaddr > UINTPTR_MAX - phdr.p_memsz)
+		return TEE_ERROR_SECURITY;
 	state->vasize = phdr.p_vaddr + phdr.p_memsz;
 
```

Comment:
I have a slight preference for Jens' solution because it does not involve UINTPTR_MAX and it is easy to read as "the value has wrapped". So, assuming it's not a problem to detect the error *after* `state->vasize` has been set (which is OK I think), then I would go for this solution.